File: mycu/core/providers/dining.py
"""Dining menus — "what's at Home Cooking today".

**This one is fully known.** Unlike the chapel provider, nothing here is a
guess: the endpoint, its parameters and its response shape were all verified
against the live service on 2026-09-16, and a real capture is committed at
``tests/fixtures/diningdata_cedarville_edu_api_menus.json``.

.. rubric:: The endpoint

    GET https://diningdata.cedarville.edu/api/menus?days=N

Unauthenticated — `diningdata.cedarville.edu`'s own page fetches it with
``credentials: "omit"``, so there is no session to have and nothing to log into.
That is why :class:`~mycu.core.transport.TransportRouter` sends this origin to
:class:`~mycu.core.transport.HttpTransport` rather than through the WebView.

Response shape, verified::

    {
      "2026-09-16": [
        {
          "venue": "Home Cooking",
          "meal":  "Breakfast",
          "slot":  "breakfast",
          "items": [
            {"name": "Chorizo Sausage Patties", "allergens": []},
            {"name": "Shredded Cheese",
             "allergens": [{"url": "…/hasDairy.png", "alt": "dairy"}]}
          ]
        },
        …
      ],
      "2026-09-17": [ … ]
    }

About twenty blocks per day. ``venue`` is the station — observed values include
Home Cooking, Grille, Italian, Habanero, SubZone, Salad & Soup, Bake Shoppe,
Allergen Aware, Garden Bites, Power Bar, Self Cook, Breakfast All Day, Self
Cook. ``meal`` is ``"Breakfast"``/``"Lunch"``/``"Dinner"`` for stations tied to
a sitting, and ``null`` for all-day stations, which carry ``slot: "anytime"``
instead.

``days=N`` was tested at 1, 7 and 14 and returns exactly N dates starting
today. The server caps it at 31: ``days=60``, ``120`` and ``365`` all come back
with 31 dates (about half a megabyte).

``start=YYYY-MM-DD`` moves the first date, in either direction. The site's own
script never sends it, but it was verified on 2026-09-22 against dates from
2025-09-01 to 2026-11-15, all of which returned real menus. It is what lets
the Chucks tab page back to last week or forward past the 31-day cap.

A date with nothing posted — a holiday, a break, or simply too far out — is
not an error. It comes back HTTP 200 with a single placeholder block::

    {"2026-12-25": [{"venue": "No Venues Found", "meal": null,
                     "slot": "anytime", "items": []}]}

which parses like any other block and is dropped by
:meth:`~mycu.core.models.DayMenu.for_venue`, leaving an empty day. There is also a ``refresh=1`` parameter, which the site's own script
uses hourly; this provider never sends it, because it appears to force an
upstream refetch from Pioneer College Caterers (``my.pcconline.com``) and there
is no reason for a personal app to make someone else's server work harder.
"""

from __future__ import annotations

import logging
from datetime import date, datetime, time
from typing import Any

from ..errors import ParseError
from ..models import HOME_COOKING, DayMenu, MenuBlock, MenuItem
from ..transport import DINING_BASE, Response
from .base import Provider

log = logging.getLogger(__name__)

#: Path including the origin — this provider does not live on Self-Service, and
#: `TransportRouter` reads the origin to pick a transport.
DINING_PATH = f"{DINING_BASE}/api/menus"

#: How far ahead to ask for. One week: enough for "what's for dinner on Friday"
#: without pulling a quarter-megabyte of JSON onto a phone.
DEFAULT_DAYS = 7


class DiningProvider(Provider[tuple[DayMenu, ...]]):
    """Menus for ``days`` days from ``start`` (default today), oldest first."""

    label = "Dining"

    def __init__(self, transport, days: int = DEFAULT_DAYS, start: date | None = None) -> None:
        super().__init__(transport)
        self.days = max(1, int(days))
        #: First date to ask for; ``None`` leaves it to the server, which means today.
        self.start = start

    @property
    def path(self) -> str:  # type: ignore[override]
        path = f"{DINING_PATH}?days={self.days}"
        if self.start is not None:
            path += f"&start={self.start.isoformat()}"
        return path

    def parse(self, response: Response) -> tuple[DayMenu, ...]:
        return parse_menus(response.json())


def parse_menus(payload: Any) -> tuple[DayMenu, ...]:
    """Turn the API's ``{date: [block]}`` mapping into :class:`DayMenu` objects.

    Sorted by date, because dict ordering is a JSON serialisation detail and
    not something a UI should depend on.

    Individual malformed blocks are skipped with a warning rather than taken as
    fatal: a menu missing one station is still a useful menu, and this endpoint
    is ultimately reformatting someone else's data.
    """
    if not isinstance(payload, dict):
        raise ParseError(
            f"expected a {{date: [blocks]}} object from the dining API, "
            f"got {type(payload).__name__}"
        )

    days: list[DayMenu] = []
    for raw_date in sorted(payload):
        on = _parse_iso_date(raw_date)
        if on is None:
            log.warning("dining: skipping unparseable date key %r", raw_date)
            continue

        raw_blocks = payload[raw_date]
        if not isinstance(raw_blocks, list):
            log.warning("dining: %s did not map to a list of blocks", raw_date)
            continue

        blocks = []
        for raw_block in raw_blocks:
            block = _parse_block(raw_block)
            if block is not None:
                blocks.append(block)

        days.append(DayMenu(on=on, blocks=tuple(blocks)))

    if not days:
        raise ParseError("the dining API returned no usable days")

    return tuple(days)


def _parse_iso_date(text: str) -> date | None:
    try:
        return datetime.strptime(str(text), "%Y-%m-%d").date()
    except (ValueError, TypeError):
        return None


def _parse_block(raw: Any) -> MenuBlock | None:
    if not isinstance(raw, dict):
        return None

    venue = _clean(raw.get("venue"))
    if not venue:
        return None

    return MenuBlock(
        venue=venue,
        # `meal` is null for all-day stations; normalise to "" so the model
        # never has to think about None, and `sort_key` puts them last.
        meal=_clean(raw.get("meal")),
        slot=_clean(raw.get("slot")),
        items=tuple(
            item for item in (_parse_item(i) for i in raw.get("items") or []) if item
        ),
    )


def _parse_item(raw: Any) -> MenuItem | None:
    if not isinstance(raw, dict):
        return None

    name = _clean(raw.get("name"))
    if not name:
        return None

    # The live API sends `allergens: [{url, alt}]`. The site's own menu.js still
    # reads `tags`, which the API no longer returns — its renderer is out of
    # date with its own backend. Accept both: if `tags` ever comes back, it
    # costs one line to keep working.
    raw_allergens = raw.get("allergens")
    if raw_allergens is None:
        raw_allergens = raw.get("tags") or []

    allergens: list[str] = []
    for entry in raw_allergens:
        if isinstance(entry, dict):
            label = _clean(entry.get("alt") or entry.get("name"))
        else:
            label = _clean(entry)
        if label and label not in allergens:
            allergens.append(label)

    return MenuItem(name=name, allergens=tuple(allergens))


def _clean(value: Any) -> str:
    """Normalise whitespace; ``None`` becomes ``""``.

    Worth doing: the upstream data has doubled spaces in real dish names
    ("Eggs with Peppers and  Onion" is in the committed fixture, verbatim).
    """
    if value is None:
        return ""
    return " ".join(str(value).split())


def home_cooking_for(days: "tuple[DayMenu, ...]", on: date | None = None) -> tuple[MenuBlock, ...]:
    """Home Cooking's breakfast, lunch and dinner for one date.

    Defaults to today. Returns an empty tuple if that date is not in the
    payload — which happens legitimately: the payload covers only the window
    that was asked for.
    """
    target = on or date.today()
    for day in days:
        if day.on == target:
            return day.for_venue(HOME_COOKING)
    return ()


#: When each sitting stops being "next".
#:
#: **These are ours, not the API's.** The menu feed carries no serving times at
#: all — only a ``slot`` label — so something has to decide when breakfast stops
#: being the meal you are about to eat. These are the posted Chuck's windows
#: rounded outward, so the answer changes a little late rather than a little
#: early: being told "up next: lunch" while you are still eating breakfast is
#: the more annoying of the two failures.
#:
#: They affect *which menu is shown first* and nothing else. The full day is on
#: the Chucks tab either way, so a wrong guess here costs a tap, not a meal.
SERVING_ENDS = {
    "breakfast": time(10, 30),
    "lunch": time(16, 0),
    "dinner": time(20, 0),
}


def next_meal_block(
    days: "tuple[DayMenu, ...]",
    now: datetime | None = None,
    venue: str = HOME_COOKING,
) -> "tuple[date, MenuBlock] | None":
    """The sitting a reader is most likely about to eat, with its date.

    Walks forward from ``now``: the first sitting today that has not finished
    serving, and failing that the first sitting on the next day in the payload.
    After dinner this means tomorrow's breakfast, which is the honest answer —
    the alternative is showing a menu for a meal that is already over.

    Returns ``None`` when nothing is left in the payload at all. Empty blocks
    are skipped rather than returned: a heading with no dishes under it reads
    as a bug, and the next real sitting is the useful answer.

    ``now`` is injectable because the whole behaviour is a function of the
    clock, and a test that could only be run before 10:30am would be no test.
    """
    moment = now or datetime.now()
    today = moment.date()

    for day in sorted(days, key=lambda d: d.on):
        if day.on < today:
            continue
        for block in day.for_venue(venue):
            # `for_venue` includes all-day stations (slot "anytime"), which are
            # never "next" — they are always on.
            ends = SERVING_ENDS.get(block.slot.casefold())
            if ends is None or not block.items:
                continue
            if day.on > today or moment.time() < ends:
                return day.on, block
    return None

File: src/core/providers/dining.cpp
#include "dining.h"

#include "../log.h"
#include "../pyjson.h"

#include <QJsonArray>
#include <QJsonObject>
#include <QRegularExpression>

#include <algorithm>

namespace mycu {

namespace {

QDate parseIsoDate(const QString &text)
{
    static const QRegularExpression shape(QStringLiteral("^(\\d{4})-(\\d{1,2})-(\\d{1,2})$"));
    const QRegularExpressionMatch m = shape.match(text);
    if (!m.hasMatch())
        return {};
    return QDate(m.captured(1).toInt(), m.captured(2).toInt(), m.captured(3).toInt());
}

std::optional<MenuItem> parseItem(const QJsonValue &raw)
{
    if (!raw.isObject())
        return std::nullopt;
    const QJsonObject obj = raw.toObject();

    MenuItem item;
    item.name = json::clean(obj.value(QStringLiteral("name")));
    if (item.name.isEmpty())
        return std::nullopt;

    // The live API sends `allergens: [{url, alt}]`. The site's own menu.js
    // still reads `tags`, which the API no longer returns — its renderer is out
    // of date with its own backend. Accept both: if `tags` ever comes back, it
    // costs one line to keep working.
    QJsonValue rawAllergens = obj.value(QStringLiteral("allergens"));
    if (rawAllergens.isUndefined() || rawAllergens.isNull())
        rawAllergens = obj.value(QStringLiteral("tags"));

    for (const QJsonValue &entry : rawAllergens.toArray()) {
        QString label;
        if (entry.isObject()) {
            const QJsonObject tag = entry.toObject();
            const QJsonValue alt = tag.value(QStringLiteral("alt"));
            label = json::clean(json::truthy(alt) ? alt : tag.value(QStringLiteral("name")));
        } else {
            label = json::clean(entry);
        }
        if (!label.isEmpty() && !item.allergens.contains(label))
            item.allergens.append(label);
    }
    return item;
}

std::optional<MenuBlock> parseBlock(const QJsonValue &raw)
{
    if (!raw.isObject())
        return std::nullopt;
    const QJsonObject obj = raw.toObject();

    MenuBlock block;
    block.venue = json::clean(obj.value(QStringLiteral("venue")));
    if (block.venue.isEmpty())
        return std::nullopt;

    // `meal` is null for all-day stations; normalised to "" so the model never
    // has to think about null, and sortKey() puts them last.
    block.meal = json::clean(obj.value(QStringLiteral("meal")));
    block.slot = json::clean(obj.value(QStringLiteral("slot")));
    for (const QJsonValue &rawItem : obj.value(QStringLiteral("items")).toArray()) {
        if (auto item = parseItem(rawItem))
            block.items.append(*item);
    }
    return block;
}

// Chuck's posted serving hours, per slot, keyed by the kind of day.
struct DayHours
{
    ServingHours breakfast, lunch, dinner;
};

const DayHours WEEKDAY{{QTime(7, 0), QTime(9, 30)},
                       {QTime(10, 30), QTime(14, 30)},
                       {QTime(16, 30), QTime(19, 30)}};
const DayHours SATURDAY{{QTime(8, 0), QTime(9, 0)},
                        {QTime(11, 0), QTime(13, 0)},
                        {QTime(16, 30), QTime(18, 30)}};
const DayHours SUNDAY{{QTime(8, 0), QTime(9, 0)},
                      {QTime(11, 30), QTime(14, 0)},
                      {QTime(17, 0), QTime(19, 30)}};

// "7am" / "9:30am" / "2:30pm".
QString clockShort(QTime at)
{
    const int hour = at.hour() % 12 == 0 ? 12 : at.hour() % 12;
    const QString minutes = at.minute() ? QStringLiteral(":%1").arg(at.minute(), 2, 10, QLatin1Char('0'))
                                        : QString();
    return QString::number(hour) + minutes + (at.hour() < 12 ? QStringLiteral("am") : QStringLiteral("pm"));
}

} // namespace

DiningProvider::DiningProvider(TransportPtr transport, int days, QDate start)
    : m_transport(std::move(transport))
    , m_days(std::max(1, days))
    , m_start(start)
{}

QString DiningProvider::path() const
{
    QString path = DINING_PATH + QStringLiteral("?days=") + QString::number(m_days);
    if (m_start.isValid())
        path += QStringLiteral("&start=") + m_start.toString(Qt::ISODate);
    return path;
}

QList<DayMenu> DiningProvider::fetch()
{
    const Response response = m_transport->get(path());
    return parse(response.raiseForSession());
}

QList<DayMenu> DiningProvider::parse(const Response &response)
{
    return parseMenus(response.json());
}

QList<DayMenu> parseMenus(const QJsonValue &payload)
{
    if (!payload.isObject()) {
        throw ParseError(QStringLiteral("expected a {date: [blocks]} object from the dining API, got %1")
                             .arg(json::typeName(payload)));
    }
    const QJsonObject obj = payload.toObject();

    QStringList keys = obj.keys();
    keys.sort();

    QList<DayMenu> days;
    for (const QString &rawDate : keys) {
        const QDate on = parseIsoDate(rawDate);
        if (!on.isValid()) {
            qCWarning(lcDining) << "dining: skipping unparseable date key" << rawDate;
            continue;
        }
        const QJsonValue rawBlocks = obj.value(rawDate);
        if (!rawBlocks.isArray()) {
            qCWarning(lcDining) << "dining:" << rawDate << "did not map to a list of blocks";
            continue;
        }

        DayMenu day{on, {}};
        for (const QJsonValue &rawBlock : rawBlocks.toArray()) {
            if (auto block = parseBlock(rawBlock))
                day.blocks.append(*block);
        }
        days.append(day);
    }

    if (days.isEmpty())
        throw ParseError(QStringLiteral("the dining API returned no usable days"));
    return days;
}

QList<MenuBlock> homeCookingFor(const QList<DayMenu> &days, QDate on)
{
    for (const DayMenu &day : days) {
        if (day.on == on)
            return day.forVenue(HOME_COOKING);
    }
    return {};
}

std::optional<ServingHours> servingHours(QDate on, const QString &slot)
{
    const int weekday = on.dayOfWeek(); // Monday = 1 … Sunday = 7
    const DayHours &hours = weekday == 6 ? SATURDAY : weekday == 7 ? SUNDAY : WEEKDAY;
    const QString key = slot.toCaseFolded();
    if (key == u"breakfast")
        return hours.breakfast;
    if (key == u"lunch")
        return hours.lunch;
    if (key == u"dinner")
        return hours.dinner;
    return std::nullopt;
}

QString formatHours(QTime start, QTime end)
{
    return clockShort(start) + QStringLiteral("–") + clockShort(end);
}

std::optional<std::pair<QDate, MenuBlock>> nextMealBlock(const QList<DayMenu> &days,
                                                         const QDateTime &now, const QString &venue)
{
    const QDate today = now.date();

    QList<DayMenu> ordered = days;
    std::stable_sort(ordered.begin(), ordered.end(),
                     [](const DayMenu &a, const DayMenu &b) { return a.on < b.on; });

    for (const DayMenu &day : ordered) {
        if (day.on < today)
            continue;
        for (const MenuBlock &block : day.forVenue(venue)) {
            // forVenue() includes all-day stations (slot "anytime"), which are
            // never "next" — they are always on.
            const std::optional<ServingHours> hours = servingHours(day.on, block.slot);
            if (!hours || block.items.isEmpty())
                continue;
            if (day.on > today || now.time() < hours->second)
                return std::make_pair(day.on, block);
        }
    }
    return std::nullopt;
}

} // namespace mycu
